File: iosApp/iosApp/ui/components/AppLoader.swift
import SwiftUI
import SharedLogic

/// The app's one coin loader, for every signed-out screen and the wizard.
///
/// Whenever it is up, it is in the middle of the screen with everything behind it
/// blurred, and it stays for at least `LoaderTiming.MIN_VISIBLE_MS` — or as long
/// as the work really takes. A card with a coin resting on its edge tells the
/// loader where that coin is, so the loader's coin sets off from there and goes
/// back there: one coin moving, rather than one vanishing while another appears.
final class AppLoader: ObservableObject {
    /// True while the loader's coin is away from its card; the card hides its own
    /// meanwhile.
    @Published fileprivate(set) var coinAway = false

    /// Where a card's coin rests, in global coordinates. Not published: it is only
    /// read as a trip starts, and publishing every layout pass would redraw the app.
    fileprivate(set) var resting: CGPoint?
    private var restingOwner: UUID?

    fileprivate func place(_ owner: UUID, at point: CGPoint) {
        restingOwner = owner
        resting = point
    }

    fileprivate func release(_ owner: UUID) {
        guard restingOwner == owner else { return }
        restingOwner = nil
        resting = nil
    }
}

private struct AppLoaderKey: EnvironmentKey {
    static let defaultValue = AppLoader()
}

extension EnvironmentValues {
    var appLoader: AppLoader {
        get { self[AppLoaderKey.self] }
        set { self[AppLoaderKey.self] = newValue }
    }
}

/// Wraps the whole app: blurs it and puts the coin over it while `active`, held
/// for the minimum however quickly the work finishes.
struct LoaderHost<Content: View>: View {
    let active: Bool
    @ObservedObject var loader: AppLoader
    @ViewBuilder var content: () -> Content

    @State private var shown = false
    @State private var shownAt = Date.distantPast
    @State private var coinVisible = false
    @State private var offset: CGSize = .zero
    @State private var scale: CGFloat = LoaderHost.loadingScale
    @State private var opacity: Double = 0

    private static var travel: Double { 0.42 }
    private static var fade: Double { 0.2 }
    /// The resting coin is 88pt; in the middle of the screen it reads at 112pt.
    private static var loadingScale: CGFloat { 112 / 88 }

    var body: some View {
        ZStack {
            content()
                .environment(\.appLoader, loader)
                .blur(radius: shown ? 18 : 0)
                .animation(.easeInOut(duration: Self.travel), value: shown)
                // Nothing behind the loader can be pressed while it is up.
                .allowsHitTesting(!shown)

            Color(.systemBackground)
                .opacity(shown ? 0.2 : 0)
                .animation(.easeInOut(duration: Self.travel), value: shown)
                .ignoresSafeArea()
                .allowsHitTesting(false)

            if coinVisible {
                CoinBadge()
                    .scaleEffect(scale)
                    .opacity(opacity)
                    .offset(offset)
                    .frame(maxWidth: .infinity, maxHeight: .infinity)
                    .ignoresSafeArea()
                    .allowsHitTesting(false)
            }
        }
        .task(id: active) { await follow(active) }
    }

    @MainActor
    private func follow(_ active: Bool) async {
        if active {
            guard !shown else { return }
            shownAt = Date()
            shown = true
            // Typing is over for now, and a keyboard would cover the coin.
            dismissKeyboard()
            await travelOut()
        } else if shown {
            let elapsed = Int64(Date().timeIntervalSince(shownAt) * 1000)
            let wait = LoaderTiming.shared.remainingMs(shownAtMs: 0, nowMs: elapsed)
            if wait > 0 { try? await Task.sleep(for: .milliseconds(wait)) }
            guard !Task.isCancelled else { return }
            shown = false
            await travelHome()
        }
    }

    /// From the card's resting coin, or from nothing, to the middle.
    @MainActor
    private func travelOut() async {
        if !coinVisible {
            if let home = homeOffset {
                offset = home
                scale = 1
                opacity = 1
            } else {
                offset = .zero
                scale = Self.loadingScale
                opacity = 0
            }
            coinVisible = true
            loader.coinAway = true
            // Let the coin appear where it starts before it is asked to move.
            try? await Task.sleep(for: .milliseconds(16))
        }
        withAnimation(.easeInOut(duration: Self.travel)) {
            offset = .zero
            scale = Self.loadingScale
        }
        withAnimation(.easeInOut(duration: Self.fade)) { opacity = 1 }
    }

    /// Back to whichever card's coin is resting now, or a fade where there is none.
    @MainActor
    private func travelHome() async {
        guard coinVisible else { return }
        if let home = homeOffset {
            withAnimation(.easeInOut(duration: Self.travel)) {
                offset = home
                scale = 1
            }
            try? await Task.sleep(for: .milliseconds(420))
        } else {
            withAnimation(.easeInOut(duration: Self.fade)) { opacity = 0 }
            try? await Task.sleep(for: .milliseconds(200))
        }
        guard !Task.isCancelled else { return }
        coinVisible = false
        loader.coinAway = false
    }

    private var homeOffset: CGSize? {
        guard let resting = loader.resting else { return nil }
        let screen = UIApplication.shared.connectedScenes
            .compactMap { $0 as? UIWindowScene }
            .first?.screen.bounds ?? UIScreen.main.bounds
        return CGSize(width: resting.x - screen.midX, height: resting.y - screen.midY)
    }
}

/// The coin resting on a card's edge. It tells the app's loader where it is, and
/// steps aside while the loader's own coin is away.
struct CardCoin: View {
    @Environment(\.appLoader) private var loader

    var body: some View {
        CardCoinBody(loader: loader)
    }
}

private struct CardCoinBody: View {
    @ObservedObject var loader: AppLoader
    @State private var owner = UUID()

    var body: some View {
        CoinBadge()
            .opacity(loader.coinAway ? 0 : 1)
            .onGeometryChange(for: CGRect.self) { $0.frame(in: .global) } action: { frame in
                loader.place(owner, at: CGPoint(x: frame.midX, y: frame.midY))
            }
            .onDisappear { loader.release(owner) }
    }
}

File: androidApp/src/main/kotlin/com/humblesolutions/finai/ui/components/AppLoader.kt
package com.humblesolutions.finai.ui.components

import android.os.Build
import android.os.SystemClock
import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.VectorConverter
import androidx.compose.animation.core.animateDpAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.offset
import androidx.compose.material3.MaterialTheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.DisposableEffect
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.Stable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableLongStateOf
import androidx.compose.runtime.mutableStateMapOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.blur
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.input.pointer.pointerInput
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalFocusManager
import androidx.compose.ui.platform.LocalSoftwareKeyboardController
import androidx.compose.ui.unit.IntSize
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.round
import com.humblesolutions.finai.util.LoaderTiming
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/**
 * The app's one coin loader, for every signed-out screen and the wizard.
 *
 * Whenever it is up, it is in the middle of the screen with everything behind it
 * blurred, and it stays for at least [LoaderTiming.MIN_VISIBLE_MS] — or as long
 * as the work really takes. A card with a coin resting on its edge tells the
 * loader where that coin is, so the loader's coin sets off from there and goes
 * back there: one coin moving, rather than one vanishing while another appears.
 */
@Stable
class AppLoader {

    /** The resting coin's centre, in window coordinates, and the card that placed it. */
    internal var resting: Offset? = null
        private set
    private var restingOwner: Any? = null

    /** True while the loader's coin is away from its card; the card hides its own meanwhile. */
    var coinAway by mutableStateOf(false)
        internal set

    private val requests = mutableStateMapOf<String, Unit>()

    /** Whether any screen outside the onboarding flow has asked for the loader. */
    val requested: Boolean get() = requests.isNotEmpty()

    fun request(key: String, active: Boolean) {
        if (active) requests[key] = Unit else requests.remove(key)
    }

    internal fun place(owner: Any, centre: Offset) {
        restingOwner = owner
        resting = centre
    }

    internal fun release(owner: Any) {
        if (restingOwner !== owner) return
        restingOwner = null
        resting = null
    }
}

val LocalAppLoader = staticCompositionLocalOf { AppLoader() }

/** Asks for the loader for as long as [active] holds and this stays on screen. */
@Composable
fun LoaderSignal(key: String, active: Boolean) {
    val loader = LocalAppLoader.current
    DisposableEffect(loader, key, active) {
        loader.request(key, active)
        onDispose { loader.request(key, false) }
    }
}

/**
 * Wraps the whole app: blurs it and puts the coin over it while [active], held
 * for the minimum however quickly the work finishes.
 */
@Composable
fun LoaderHost(active: Boolean, loader: AppLoader, content: @Composable () -> Unit) {
    val shown = rememberHeldVisible(active)
    val blur by animateDpAsState(if (shown) BlurRadius else 0.dp, tween(TRAVEL_MS), label = "loaderBlur")
    val veil by animateFloatAsState(if (shown) 1f else 0f, tween(TRAVEL_MS), label = "loaderVeil")
    var host by remember { mutableStateOf(IntSize.Zero) }

    val keyboard = LocalSoftwareKeyboardController.current
    val focus = LocalFocusManager.current
    LaunchedEffect(shown) {
        // Typing is over for now, and a keyboard would cover the coin.
        if (shown) {
            keyboard?.hide()
            focus.clearFocus()
        }
    }

    Box(Modifier.fillMaxSize().onSizeChanged { host = it }) {
        // Blur needs API 31. Below it the veil is thicker, so the screen still
        // steps back behind the coin.
        Box(Modifier.fillMaxSize().blur(blur)) {
            CompositionLocalProvider(LocalAppLoader provides loader) { content() }
        }
        if (veil > 0f) {
            val ground = MaterialTheme.colorScheme.background
            Box(
                Modifier
                    .fillMaxSize()
                    .alpha(veil)
                    .background(ground.copy(alpha = if (Build.VERSION.SDK_INT >= 31) 0.2f else 0.7f))
                    // Nothing behind the loader can be pressed while it is up.
                    .pointerInput(Unit) {
                        awaitPointerEventScope {
                            while (true) awaitPointerEvent().changes.forEach { it.consume() }
                        }
                    },
            )
        }
        TravellingCoin(shown, loader, host)
    }
}

/**
 * [active], but once it turns true it stays true for at least the minimum,
 * measured from the moment it first showed.
 */
@Composable
fun rememberHeldVisible(active: Boolean): Boolean {
    var shown by remember { mutableStateOf(active) }
    var shownAt by remember { mutableLongStateOf(SystemClock.uptimeMillis()) }
    LaunchedEffect(active) {
        if (active) {
            if (!shown) {
                shownAt = SystemClock.uptimeMillis()
                shown = true
            }
        } else if (shown) {
            delay(LoaderTiming.remainingMs(shownAt, SystemClock.uptimeMillis()))
            shown = false
        }
    }
    return shown
}

@Composable
private fun BoxScope.TravellingCoin(shown: Boolean, loader: AppLoader, host: IntSize) {
    val offset = remember { Animatable(Offset.Zero, Offset.VectorConverter) }
    val scale = remember { Animatable(1f) }
    val opacity = remember { Animatable(0f) }
    var visible by remember { mutableStateOf(false) }

    LaunchedEffect(shown) {
        val centre = Offset(host.width / 2f, host.height / 2f)
        val home = loader.resting?.takeIf { host != IntSize.Zero }?.minus(centre)
        if (shown) {
            if (!visible) {
                offset.snapTo(home ?: Offset.Zero)
                scale.snapTo(if (home != null) 1f else LOADING_SCALE)
                opacity.snapTo(if (home != null) 1f else 0f)
                visible = true
                loader.coinAway = true
            }
            coroutineScope {
                launch { offset.animateTo(Offset.Zero, tween(TRAVEL_MS)) }
                launch { scale.animateTo(LOADING_SCALE, tween(TRAVEL_MS)) }
                launch { opacity.animateTo(1f, tween(FADE_MS)) }
            }
        } else if (visible) {
            if (home != null) {
                coroutineScope {
                    launch { offset.animateTo(home, tween(TRAVEL_MS)) }
                    launch { scale.animateTo(1f, tween(TRAVEL_MS)) }
                }
            } else {
                opacity.animateTo(0f, tween(FADE_MS))
            }
            visible = false
            loader.coinAway = false
        }
    }

    if (visible) {
        CoinBadge(
            Modifier
                .align(Alignment.Center)
                .offset { offset.value.round() }
                .graphicsLayer {
                    scaleX = scale.value
                    scaleY = scale.value
                    alpha = opacity.value
                },
        )
    }
}

private val BlurRadius = 18.dp
private const val TRAVEL_MS = 420
private const val FADE_MS = 200

/** The resting coin is 88dp; in the middle of the screen it reads at 112dp. */
private const val LOADING_SCALE = 112f / 88f
